### PR TITLE
Assignment4 Matthew adams and Rachael DeBoer

### PR DESCRIPTION
**Pull request by MattA41**
@MattA41 and @rkdeboer

### CODE REVIEW

**Comment by montalcy** [comment]
Reviewing

File: Logger.Tests/EntityTests.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        FullName testName = new("Cotton", "Joe", "Eye");
+
+
+        Student employee = new(testName);
+
+        Assert.Equal(nameTest, employee.Name);
+
+    }
+    [Fact]
+    public void Book_NameMatches()
```
Comment (montalcy) [inline Logger.Tests/EntityTests.cs:39]: I appreciate ya'll making a whole new test class, I put it all in one, but this seems a lot cleaner.

File: Logger.Tests/FullNameTests.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        Assert.Equal(nameTest, name.Name);
+    }
+    [Fact] 
+    public void NoFirstName_ThrowsException()
+    {
+        
+        Assert.Throws<ArgumentNullException>(() => new FullName(null!, "Joe", "Eye"));
+    }
+    [Fact]
+    public void NoLastName_ThrowsException()
```
Comment (montalcy) [inline Logger.Tests/FullNameTests.cs:34]: Maybe this could all be with EntityTest file since they all seem to be taking care of the name equality

Comment (montalcy) [inline Logger.Tests/FullNameTests.cs:34]: Good job catching the null exception!

File: Logger.Tests/StorageTests.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+        storage.Add(student);
+
+        Assert.True(storage.Contains(student));
+        storage.Remove(student);
+        Assert.False(storage.Contains(student));
+    }
+    [Fact]
+
+    public void RemoveBook_BookRemoved()
```
Comment (montalcy) [inline Logger.Tests/StorageTests.cs:75]: I forgot to test removed. Good catch!

File: Logger/Employee.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+public record Employee : Person
+{
+
+    public Employee(FullName name) : base(name) { }
```
Comment (montalcy) [inline Logger/Employee.cs:11]: Do we not need to do the Guid?

File: Logger/FullName.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+    public class FullName(string firstName, string lastName, string? middleName = null)
+    {
+        public string FirstName { get; } = string.IsNullOrWhiteSpace(firstName) ? throw new ArgumentNullException(nameof(firstName)) : firstName;
+        public string LastName { get; } = string.IsNullOrWhiteSpace(lastName) ? throw new ArgumentNullException(nameof(lastName)) : lastName;
+        public string? MiddleName { get; } = middleName;
+        public string Name => (MiddleName == null ? $"{FirstName} {LastName}" : $"{FirstName} {MiddleName} {LastName}");
```
Comment (montalcy) [inline Logger/FullName.cs:13]: Good efficient use of your code

File: Logger/FullName.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+    public class FullName(string firstName, string lastName, string? middleName = null)
```
Comment (montalcy) [inline Logger/FullName.cs:8]: make sure to add your comments on nullability and explicit/implicit

File: Logger/Person.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+using System.Linq;
+using System.Reflection.Metadata.Ecma335;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+    public record class Person(FullName FullName) : BaseEntity
+    {
+        
+        public override string Name { get => FullName.Name; }
```
Comment (montalcy) [inline Logger/Person.cs:12]: I didnt know you could do the getter this way, nice.

File: Logger/Employee.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+public record Employee : Person
```
Comment (montalcy) [inline Logger/Employee.cs:8]: Nice implementation of person record class to simplify the employee and student

File: Logger.Tests/EntityTests.cs (hunk 1 above)
Comment (rkdeboer) [reply Logger.Tests/EntityTests.cs:39]: Thank you! I think keeping them seprate is the key to organization

File: Logger.Tests/FullNameTests.cs (hunk 2 above)
Comment (rkdeboer) [reply Logger.Tests/FullNameTests.cs:34]: While I understand this, I think the tests should remain seprate as one checks to see if the name in FullName is set and the other tests checks to see if the FullName is applied to entities. Good observation though

Comment (rkdeboer) [reply Logger.Tests/FullNameTests.cs:34]: Thank you!

File: Logger.Tests/StorageTests.cs (hunk 3 above)
Comment (rkdeboer) [reply Logger.Tests/StorageTests.cs:75]: Thank you!

File: Logger/FullName.cs (hunk 5 above)
Comment (rkdeboer) [reply Logger/FullName.cs:13]: Thanks!

File: Logger/FullName.cs (hunk 6 above)
Comment (rkdeboer) [reply Logger/FullName.cs:8]: Good catch I will add those comments!

File: Logger/Person.cs (hunk 7 above)
Comment (rkdeboer) [reply Logger/Person.cs:12]: Thank you!

File: Logger/Employee.cs (hunk 8 above)
Comment (rkdeboer) [reply Logger/Employee.cs:8]: Thank you!

File: Logger/Employee.cs (hunk 4 above)
Comment (rkdeboer) [reply Logger/Employee.cs]: I missed that thank you!

File: Logger/Book.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+using System.Linq;
+using System.Reflection.Metadata.Ecma335;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Logger;
+public record Book(string Title, Guid Id)  : BaseEntity
+{
+    public override string Name { get; } = Title;
+    private readonly Guid id = Id;
```
Comment (BenjaminMichaelis) [inline Logger/Book.cs:12]:
```
    private readonly Guid id = Id;
```
This is no the correct naming. Also, this should be a property, as discussed previously, this should not be a field

File: Logger/IEntity.cs [hunk 10]
```diff
@@ -2,6 +2,7 @@
 public interface IEntity
 {
 
-        // Place members here.
+    Guid Id { get; init; }
```
Comment (BenjaminMichaelis) [inline Logger/IEntity.cs:5]: Why do you need an `init`?

File: Logger/Book.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection.Metadata.Ecma335;
```
Comment (BenjaminMichaelis) [inline Logger/Book.cs:4]: What is this used for?

File: Logger.Tests/EntityTests.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Logger.Tests;
+
+public class EntityTests
+{
+    [Fact]
+    public void Student_GivenName_NameMatches()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/EntityTests.cs:13]: All of these entities (Student, Employee, etc) should be in different files.

A model of tests should be in their own files

**Pull request closed without merging**